File: projects/spring-boot-restv2/demo/src/test/java/com/sourabh/demo/servicetest/CancellationServiceTest.java
package com.sourabh.demo.servicetest;


import com.sourabh.demo.model.Customer;
import com.sourabh.demo.model.Room;
import com.sourabh.demo.model.Transaction;
import com.sourabh.demo.repository.CustomerRepository;
import com.sourabh.demo.service.CancellationService;
import com.sourabh.demo.service.RoomService;
import com.sourabh.demo.service.TransactionService;
import org.json.JSONObject;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.mock.mockito.MockBean;


import java.time.LocalDateTime;

import static org.mockito.Mockito.when;

@SpringBootTest
class CancellationServiceTest {
    @InjectMocks
    CancellationService cancellationService;
    @Mock
    CustomerRepository customerRepository;
    @Mock
    RoomService roomService;
    @Mock
    TransactionService transactionService;



    @Test
    void getCustomerTest(){
        Customer customer = new Customer(1,"Sourabh","Suit",6,6000,true,"UPI", LocalDateTime.now(),true);
        when(customerRepository.findById(1)).thenReturn(java.util.Optional.of(customer));
        Assertions.assertEquals(customer, cancellationService.getCustomer(1));
    }
    @Test
    void cancelRoomTest(){
        Customer customer = new Customer(1,"Sourabh","Suit",6,6000,true,"UPI", LocalDateTime.now(),true);
        when(customerRepository.findById(1)).thenReturn(java.util.Optional.of(customer));

        Room room = new Room(1,"Suit",1200,12,true);
        when(roomService.findByType("Suit")).thenReturn(room);

        Transaction transaction = new Transaction();
        transaction.setId(1);
        transaction.setCustomer(customer);
        transaction.setAmount(6000);
        when(transactionService.getById(1)).thenReturn(transaction);

        String actual=cancellationService.cancelRoom(1);
        JSONObject obj = new JSONObject(actual);
        Assertions.assertEquals(4800,obj.get("Refund"));

    }

    @Test
    void cancelRoomTest2(){
        Customer customer = new Customer(1,"Sourabh","Suit",6,6000,false,"UPI", LocalDateTime.now(),true);
        when(customerRepository.findById(1)).thenReturn(java.util.Optional.of(customer));

        Room room = new Room(1,"Suit",1200,12,true);
        when(roomService.findByType("Suit")).thenReturn(room);

        Transaction transaction = new Transaction();
        transaction.setId(1);
        transaction.setCustomer(customer);
        transaction.setAmount(6000);
        when(transactionService.getById(1)).thenReturn(transaction);

        String actual=cancellationService.cancelRoom(1);
        JSONObject obj = new JSONObject(actual);
        Assertions.assertEquals(0,obj.get("Refund"));

    }



}

File: projects/spring-boot-rest-kt/hotel-reservation/src/test/kotlin/com/sourabh/hotelreservation/servicetest/CancellationServiceTest.kt
package com.sourabh.hotelreservation.servicetest

import com.sourabh.hotelreservation.model.Customer
import com.sourabh.hotelreservation.model.Room
import com.sourabh.hotelreservation.model.Transaction
import com.sourabh.hotelreservation.repository.CustomerRepository
import com.sourabh.hotelreservation.service.CancellationService
import com.sourabh.hotelreservation.service.RoomService
import com.sourabh.hotelreservation.service.TransactionService
import org.json.JSONObject
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import org.mockito.InjectMocks
import org.mockito.Mock
import org.mockito.Mockito
import org.springframework.boot.test.context.SpringBootTest
import java.time.LocalDateTime
import java.util.*

@SpringBootTest
class CancellationServiceTest {
    @InjectMocks
    var cancellationService: CancellationService? = null

    @Mock
    var customerRepository: CustomerRepository? = null

    @Mock
    var roomService: RoomService? = null

    @Mock
    var transactionService: TransactionService? = null


    @Test
    fun getCustomerTest() {
        val customer = Customer(1, "Sourabh", "Suit", 6, 6000F, true, LocalDateTime.now(), "UPI", true)
        Mockito.`when`(customerRepository!!.findById(1)).thenReturn(Optional.of(customer))
        Assertions.assertEquals(customer, cancellationService!!.getCustomer(1))
    }

    @Test
    fun cancelRoomTest() {
        val customer = Customer(1, "Sourabh", "Suit", 6, 6000F, true, LocalDateTime.now(), "UPI", true)
        Mockito.`when`(customerRepository!!.findById(1)).thenReturn(Optional.of(customer))

        val room = Room(1, "Suit", 1200F, 12, true)
        Mockito.`when`(roomService!!.findByType("Suit")).thenReturn(room)

        val transaction = Transaction(1, customer, 6000F)
        Mockito.`when`(transactionService!!.getById(1)).thenReturn(transaction)
        val actual = cancellationService!!.cancelRoom(1)
        val obj = JSONObject(actual)
        Assertions.assertEquals(4800, obj["Refund"])
    }
    @Test
    fun cancelRoomTest2() {
        val customer = Customer(1, "Sourabh", "Suit", 6, 6000F, false, LocalDateTime.now(), "UPI", true)
        Mockito.`when`(customerRepository!!.findById(1)).thenReturn(Optional.of(customer))

        val room = Room(1, "Suit", 1200F, 12, true)
        Mockito.`when`(roomService!!.findByType("Suit")).thenReturn(room)

        val transaction = Transaction(1, customer, 6000F)
        Mockito.`when`(transactionService!!.getById(1)).thenReturn(transaction)

        val actual = cancellationService!!.cancelRoom(1)
        val obj = JSONObject(actual)
        Assertions.assertEquals(0, obj["Refund"])
    }



}